scipy/sparse/linalg/isolve/lsqr.py

```diff
@@ -298,14 +302,17 @@ def lsqr(A, b, damp=0.0, atol=1e-8, btol=1e-8, conlim=1e8,
 
     """
     Set up the first vectors u and v for the bidiagonalization.
-    These satisfy  beta*u = b,  alfa*v = A'u.
+    These satisfy  beta*u = b - A*x,  alfa*v = A'*u.
     """
-    v = np.zeros(n)
     u = b
-    x = np.zeros(n)
+    if x0 is None:
+        x = np.zeros(n)
+    else:
+        x = np.asarray(x0)
+        u -= A.matvec(x)
```

Comment:
I was retaining the same behavior in the function as it stands in master, though it would be better to make a copy.